### PR TITLE
Update template sharepoint-adfs

### PR DESCRIPTION
**Pull request by Yvand**
# PR Checklist

Check these items before submitting a PR... 

[Contribution Guide](https://github.com/Azure/azure-quickstart-templates/blob/master/1-CONTRIBUTION-GUIDE/README.md)

[Best Practice Guide](https://github.com/Azure/azure-quickstart-templates/blob/master/1-CONTRIBUTION-GUIDE/best-practices.md)

- [x] - Please check this box once you've submitted the PR if you've read through the Contribution Guide and best practices checklist.

## Changelog

* Update SQL to SQL Server 2019 on Windows Server 2019
* Add a network security group to Azure Bastion subnet
* Rename some resources and variables with more meaningful names
* Update apiVersion of each resource to latest version
* Update DSC module NetworkingDsc from 7.4 to 8.0
* Update DSC module xPSDesiredStateConfiguration from 8.10 to 9.1
* Update DSC module ActiveDirectoryCSDsc from 4.1 to 5.0
* Update DSC module xDnsServer from 1.15 to 1.16
* Update DSC module ComputerManagementDsc from 7.0 to 8.3
* Update DSC module SqlServerDsc from 13.2 to 14.1
* Update DSC module xWebAdministration from 2.8 to 3.1.1
* Update DSC module SharePointDSC from 3.6 to 4.2

### CODE REVIEW

**Comment by Yvand** [comment]
I see that the deployment in AzQuickStarts-Public failed with this error:

```
VM has reported a failure when processing extension 'ConfigureDCVM'. Error message: "DSC Configuration 'ConfigureDCVM' completed with error(s). Following are the first few: WinRM cannot process the request. The following error with errorcode 0x80090350 occurred while using Negotiate authentication: An unknown security error occurred.
```

I never ever got this error before. Can you start a new deployment?

**Comment by Yvand** [comment]
@bmoore-msft I see the error is exactly the same but I have no idea why it occurs.
I carefully tested my changes and this error never occurred.
Is there any chance something is wrong on the CI side? Maybe with the handling of generated passwords from azuredeploy.parameters.json?

**Comment by bmoore-msft** [comment]
Not something I've seen before... but my guess is related to one of the changes.  What's in the repo now, passes the automated test (I just ran another yesterday to be sure).

**Comment by Yvand** [comment]
The error occurs because I updated the apiVersion of resources ["Microsoft.Compute/virtualMachines/extensions"](https://docs.microsoft.com/en-us/azure/templates/microsoft.compute/virtualmachines/extensions) to latest version "2019-03-01".
It works fine if I revert to "2019-12-01".
Should I leave it like this?

**Comment by Yvand** [comment]
Well I'm confused now
https://github.com/Azure/azure-quickstart-templates/pull/7858/commits/1c3a5cd01b5f2f65a76e7352acd04879f038aa0d passed ConfigureDCVM fine, and failed at the next "Microsoft.Compute/virtualMachines/extensions" resource, which is expected because I only updated ConfigureDCVM.
Now I applied the same change to all "Microsoft.Compute/virtualMachines/extensions" resources, and https://github.com/Azure/azure-quickstart-templates/pull/7858/commits/5e000399a0a822fad88b47bf1b867119fbcdafef fails again at ConfigureDCVM...
In the [template folder](https://github.com/Azure/azure-quickstart-templates/tree/master/sharepoint-adfs) I see the date changed only for Azure US Gov, did the test run also for Azure Public?

**Comment by bmoore-msft** [comment]
So 2019-12-01 works in AzureGov but not in Public Azure???  Next thing I might suggest, if you've narrowed it down to that, is to try apiVersion 2020-06-01.  Failing that - rollback to the one that works.

re: the badges - that's the test that's run on master, I just triggered Gov not public.  I'll trigger one for public since this is so much fun ;)

**Comment by Yvand** [comment]
So, it's getting more and more surprising.
Following your suggestion I updated apiVersion of all "Microsoft.Compute/virtualMachines/extensions" to "2020-06-01".
Now, ConfigureDCVM (the first extension) succeeds in both environments:

* [Permalink](https://dev.azure.com/azurequickstarts/azure-quickstart-templates/_build/results?buildId=29514&view=logs&j=275f1d19-1bd8-5591-b06b-07d489ea915a&t=7bf03a3b-69d5-5660-244f-21e634d6903f&l=471) in AzQuickStarts-Public
* [Permalink](https://dev.azure.com/azurequickstarts/azure-quickstart-templates/_build/results?buildId=29515&view=logs&j=275f1d19-1bd8-5591-b06b-07d489ea915a&t=7bf03a3b-69d5-5660-244f-21e634d6903f&l=273) in AzQuickStarts-USGov

But now, both fail with the same error, at a different stage:
"Status Message: The resource group 'azdo-777d3c39-2fb2-4655-a4e2-d1df5b497590' is in deprovisioning state and cannot perform this operation. (Code:ResourceGroupBeingDeleted)"
In AzQuickStarts-USGov it occurred 18 minutes after deployment started so it doesn't look like a timeout.
But in AzQuickStarts-Public it occurred 1h50 after deployment started.

Also, deployment in Azure Public failed for the [current version](https://github.com/Azure/azure-quickstart-templates/tree/master/sharepoint-adfs). Could it not be an issue with the CI system in Azure Public after all?

**Comment by Yvand** [comment]
Except with apiVersion 2020-06-01, in all tests CI AzQuickStarts-Public consistently fails with the same error at ConfigureDCVM.
I cannot check the logs for the [current version](https://github.com/Azure/azure-quickstart-templates/tree/master/sharepoint-adfs) but I strongly suspect it failed with the same error.
Since deployment is working fine otherwise (as per my own tests and in AzQuickStarts-USGov), I think there is something wrong on CI AzQuickStarts-Public.
What do you think?

**Comment by Yvand** [comment]
@bmoore-msft please don't approve this PR yet, I think the template is valid, but maybe not.
Is it possible to let me check the logs of the failed deployment of AzQuickStarts-Public for [current version](https://github.com/Azure/azure-quickstart-templates/tree/master/sharepoint-adfs/)?

**Comment by bmoore-msft** [comment]
The test results from master:
https://dev.azure.com/azurequickstarts/azure-quickstart-templates/_build/results?buildId=29311&view=results

https://dev.azure.com/azurequickstarts/azure-quickstart-templates/_build/results?buildId=29311&view=results

Looks similar, USGov passed, Public failed

**Comment by Yvand** [comment]
@bmoore-msft I think there was a copy/paste issue in your links, they are identical and point to the run on AzQuickStarts-USGov.

I observe a somewhat similar behavior with this [DTL template](https://github.com/Azure/azure-devtestlab/tree/master/Environments/SharePoint-AllVersions), and I have good reasons (I can provide more details if you wish) to believe it's caused by the region.
Is it possible to test AzQuickStarts-Public on region "East US 2" (either the PR or the current version)? I'm confident it will complete successfully there.

**Comment by bmoore-msft** [comment]
ok, updated the link, sorry 'bout that...

You can set the location param value in the param file, that should move everything to the new region.  The RG will still be westus, but the resources will follow the location param.  You may run into quota issues but can try...

**Comment by bmoore-msft** [comment]
oh, yes and more details if we figure this out... if it's a platform issue, we're not likely the only ones experiencing it ;)

**Comment by Yvand** [comment]
AzQuickStarts-Public completes successfully if it deploys to "East US 2", so I'm convinced it depends on the region.

Initially, this PR always failed in AzQuickStarts-Public at ConfigureDCVM with a WinRM error (see my first message).
Now this error is completely gone (no idea why), but it fails with a time out error on ConfigureSQLVM and ConfigureSPVM (both depend on ConfigureDCVM and start only if it succeeded).

I bet the cause of the time out is the same as for this [DTL template](https://github.com/Azure/azure-devtestlab/tree/master/Environments/SharePoint-AllVersions):
Azure extension Microsoft.Powershell.DSC does not resume after the VMs SQL and SP reboot.
I found a simple workaround: Run this PowerShell command on any server that "hangs":
`C:\Packages\Plugins\Microsoft.Powershell.DSC\2.80.0.3\bin\enable.ps1 -StartAsyncProcess`
It forces the Azure extension Microsoft.Powershell.DSC to resume, and communicates success to Azure once DSC completes.

I'm pretty sure this issue is not specific to this template, it impacts any that uses DSC and needs reboots (the current version of this template used to work and now it fails as well).
Ideally, the team that develops Azure extension Microsoft.Powershell.DSC should look at this, but I don't know how to reach them.

**Comment by Yvand** [comment]
@bmoore-msft do you have a comment/suggestion on my previous message?
Should I remove the location parameter so you can approve this PR?

**Comment by Yvand** [comment]
@bmoore-msft finally I found a solution to this time out issue.
Somehow, it happens on DSC extensions that do not start immediately after their VM is created (they have other dependencies).
I updated the template so that now, all Azure DSC extension depend only on their own VM.

PR is ready to be approved on my side.
I hope you can approve it soon as the current version of the template fails due to that time out

File: sharepoint-adfs/dsc/ConfigureDCVM.ps1 [hunk 1]
```diff
@@ -280,14 +280,14 @@
             ClaimsProviderName = @("Active Directory")
             WsFederationEndpoint = "https://$SPTrustedSitesName.$DomainFQDN/_trust/"
             AdditionalWSFedEndpoint = @("https://*.$DomainFQDN/")
-            IssuanceAuthorizationRules = '=> issue(Type = "https://schemas.microsoft.com/authorization/claims/permit", value = "true");'
+            IssuanceAuthorizationRules = '=> issue(Type = "http://schemas.microsoft.com/authorization/claims/permit", value = "true");'
```
Comment (bmoore-msft) [inline sharepoint-adfs/dsc/ConfigureDCVM.ps1:283]: why not https?

File: sharepoint-adfs/dsc/ConfigureDCVM.ps1 [hunk 2]
```diff
@@ -280,14 +280,14 @@
...
             IssuanceTransformRules = @"
 @RuleTemplate = "LdapClaims"
 @RuleName = "AD"
-c:[Type == "https://schemas.microsoft.com/ws/2008/06/identity/claims/windowsaccountname", Issuer == "AD AUTHORITY"]
+c:[Type == "http://schemas.microsoft.com/ws/2008/06/identity/claims/windowsaccountname", Issuer == "AD AUTHORITY"]
```
Comment (bmoore-msft) [inline sharepoint-adfs/dsc/ConfigureDCVM.ps1:287]: why not https?

File: sharepoint-adfs/dsc/ConfigureDCVM.ps1 [hunk 3]
```diff
@@ -280,14 +280,14 @@
...
 => issue(
 store = "Active Directory", 
-types = ("https://schemas.xmlsoap.org/ws/2005/05/identity/claims/emailaddress", "https://schemas.microsoft.com/ws/2008/06/identity/claims/role"), 
+types = ("http://schemas.xmlsoap.org/ws/2005/05/identity/claims/emailaddress", "http://schemas.microsoft.com/ws/2008/06/identity/claims/role"),
```
Comment (bmoore-msft) [inline sharepoint-adfs/dsc/ConfigureDCVM.ps1:290]: why not https?

File: sharepoint-adfs/dsc/ConfigureSPVM.ps1 [hunk 4]
```diff
@@ -508,15 +508,15 @@ configuration ConfigureSPVM
             Description                  = "Federation with $DomainFQDN"
             Realm                        = "https://$SPTrustedSitesName.$DomainFQDN"
             SignInUrl                    = "https://adfs.$DomainFQDN/adfs/ls/"
-            IdentifierClaim              = "https://schemas.xmlsoap.org/ws/2005/05/identity/claims/emailaddress"
+            IdentifierClaim              = "http://schemas.xmlsoap.org/ws/2005/05/identity/claims/emailaddress"
```
Comment (bmoore-msft) [inline sharepoint-adfs/dsc/ConfigureSPVM.ps1:511]: https? (won't comment on the rest)

File: sharepoint-adfs/dsc/ConfigureDCVM.ps1 (hunk 1 above)
Comment (Yvand) [reply sharepoint-adfs/dsc/ConfigureDCVM.ps1:283]: All those URI are claim types that are set with http in both SharePoint and [ADFS](https://docs.microsoft.com/en-us/windows-server/identity/ad-fs/technical-reference/the-role-of-claims).
To my knowledge it's just a standard / naming convention to always uses http to define a claim type.

File: sharepoint-adfs/dsc/ConfigureDCVM.ps1 (hunk 3 above)
Comment (Yvand) [reply sharepoint-adfs/dsc/ConfigureDCVM.ps1:290]: same as above

File: sharepoint-adfs/dsc/ConfigureSPVM.ps1 (hunk 4 above)
Comment (Yvand) [reply sharepoint-adfs/dsc/ConfigureSPVM.ps1:511]: same as above

**Comment by Yvand** [comment]
I found out that, somehow, the time out occurs because Azure DSC extension "just" does not resume DSC after the VM reboots.
Somehow, it only happens to DSC extensions that have multiple dependencies in the ARM template.

So in commit https://github.com/Azure/azure-quickstart-templates/pull/7858/commits/d71bf7b04a72b3935eb22bbfbe50100b17ab5ce2 I only removed the extra dependencies of resources "Microsoft.Compute/virtualMachines/extensions", so that now they only depend on their own VM.

I put those dependencies for a good reason and with this quick update deployment may fail sometimes.
I'm updating DSC on all VMs to make deployment reliable again, but at least with this PR it will work most of the time, compared to never with currently version.

File: sharepoint-adfs/dsc/ConfigureDCVM.ps1 (hunk 1 above)
Comment (bmoore-msft) [reply sharepoint-adfs/dsc/ConfigureDCVM.ps1:283]: Prior versions worked with HTTPS, any particular reason for the change?

Comment (Yvand) [reply sharepoint-adfs/dsc/ConfigureDCVM.ps1]: This one is a mistake that I totally missed, it didn't have any impact but [it should definitely be https](https://docs.microsoft.com/en-us/windows-server/identity/ad-fs/technical-reference/when-to-use-an-authorization-claim-rule), thank you for pointing it out!

**Comment by Yvand** [comment]
@bmoore-msft is this PR still in your radar?
Deployment in "AzQuickStarts-Public" failed because it was interrupted (it's what I see in the logs), there is no actual error in the template.

FYI, soon I'll submit a new PR with a big update on the whole DSC but it's not ready yet,
In the meantime, this PR should be merged ASAP as current template always times out.

**Comment by Yvand** [comment]
@bmoore-msft, my commits from today makes DSC significantly more reliable after the timeout bug.
PR is ready to be approved on my end. Are there things to validate on your side?

**Comment by bmoore-msft** [comment]
looks good...

**Pull request merged**